### PR TITLE
Layout subset matching

### PR DESCRIPTION
**Pull request by un-tone**
For https://github.com/Starcounter/RebelsLounge/issues/179.
Related PR: https://github.com/Starcounter/level1/pull/4309.

### CODE REVIEW

**Comment by warpech** [comment]
I am very excited about this! I love the tests!

So far, I have a question about the data model. If you agree with it, it will require quite substantial changes, so I will wait for your response in https://github.com/Starcounter/level1/pull/4309.

**Comment by un-tone** [comment]
Fixed.

**Comment by warpech** [comment]
I can successfully reproduce "Test 2", great! 

But "Test 1" fails at the part in bold:

> - Verify that the composition changes are preserved and **Timeline is rendered at the bottom**

Can you please investigate?

**Comment by un-tone** [comment]
Oh, I missed that case. So I should find a way how to add a Timeline view to the preserved composition. Any suggestions?

**Comment by warpech** [comment]
We need to append the `declarative-shadow-dom` part from the newly added app at the bottom of the composition restored on-the-fly (without saving).

The server does not parse HTML so it needs to be done on the client.

**Comment by un-tone** [comment]
Yes, please :-)

**Comment by warpech** [comment]
I am working on this. 

@un-tone could you please:

1. Check if [my comments in the test helper](https://github.com/Starcounter/CompositionProvider/pull/15/commits/8b85b45a8c689086f59c76aed7ca73fadebcb192) are correct
2. Change [Program.cs](https://github.com/Starcounter/CompositionProvider/blob/8b85b45a8c689086f59c76aed7ca73fadebcb192/src/CompositionProvider/Program.cs) to be backwards compatible in case an exact `HtmlViewComposition` is not defined but a legacy `HTMLComposition` is defined. This can be tested by the below "Test 3":

## Test 3

0. Wipe out the database
1. Start CompositionProvider and the `develop` branches of [inside.starcounter.com](https://github.com/Starcounter/inside.starcounter.com), Images, People, SignIn
2. Create an admin user and sign in
3. Go to http://localhost:/8080/people, click on editing a person and verify that this composition is applied: https://github.com/Starcounter/inside.starcounter.com/blob/develop/src/InsideStarcounterMapper/Content/HtmlCompositions/People-PersonPage_Images.html

**Comment by warpech** [comment]
@un-tone I have one more request: Currently, the `Version` property in `HtmlViewComposition` is unused.

What we should have is:

- every `HtmlViewComposition` instance is treated as readonly
- every change to a composition is inserted to the db as a new `HtmlViewComposition` with bumped `Version`

**Comment by un-tone** [comment]
Added backward compat.

**Comment by warpech** [comment]
@un-tone Any idea why the tests of TC are failing? Is there something wrong with `tools/Teamcity.ps1`?

**Comment by un-tone** [comment]
Implemented composition versioning with some changes in the data model: https://github.com/Starcounter/level1/pull/4322.

**Comment by un-tone** [comment]
Something strange with with TC builds. Currently it fails on such code:

```csharp
if (!(Session.Ensure().Store[nameof(MasterPage)] is MasterPage master))
```
http://teamcity/viewLog.html?tab=buildLog&logTab=tree&filter=debug&expand=all&buildId=36082#_focus=101

**Comment by warpech** [comment]
I have pushed my changes for complementing the composition with the default composition from the HTML. **Test 1** is now passing.

@un-tone, @Joltrix could you pls test? It works for me with Starcounter 2.3.1.7654

**Comment by un-tone** [comment]
There is an issue when saving a composition. It calls `LayoutInfo.Handle(Input.Composition input)` twice so we're getting duplicates.
@warpech could you check it, I guess the problem is in starcounter-include

**Comment by warpech** [comment]
Thanks for catching this! Unfortunately, this is needed because of a bug in Palindrom. See https://github.com/Starcounter/CompositionProvider/blob/360b6350688dc722e2a6607262dbb8b96be09937/src/CompositionProvider/wwwroot/sys/starcounter-include/starcounter-include.html#L360 for an explanation.

I will make a proper issue in Palindrom.

**Comment by un-tone** [comment]
I missed one thing, sorry. When it calls `LayoutInfo.Handle(Input.Composition input)` first time it sends correct html, but in second time it sends old html value. It is not reproducible when saving composition first time.

**Comment by warpech** [comment]
Thanks a lot for catching this! Can you please re-test with the submitted fix?

**Comment by un-tone** [comment]
All the problems is gone. Now it calls `LayoutInfo.Handle(Input.Composition input)` only once with correct composition value. Thanks!

**Comment by un-tone** [comment]
So currently we have issues only with TC builds.

**Comment by un-tone** [comment]
TeamCity uses MSBuild ToolsVersion 14.0 while should uses 15.0. @ubbeK could you please fix the CompositionProvider build settings?

**Comment by un-tone** [comment]
Ping @ubbeK

**Comment by warpech** [comment]
There is still some problem with composition editing. @alshakero please help to investigate:

1. Create a new empty database
2. Clone, build and run CompositionEditor `develop` branch using `run.bat` (this will start CompositionEditor)
3. Clone, build and run CompositionProvider `layout-subset` branch using `test.bat` (this will start CompositionProvider, CompositionProvider_TestApp1, CompositionProvider_TestApp2, CompositionProvider_TestHelper)
4. Clone starcounter-include `subset-composition` branch and use `starcounter-include.html` from that branch
5. Use a build of Palindrom that contains https://github.com/Palindrom/JSONPatcherProxy/pull/22
6. Go to http://localhost:8080/CompositionProvider_TestHelper/resetdata to remove all compositions
7. Go to http://172.20.10.5:8080/CompositionProvider_TestHelper/main
8. Press CTRL+E to open composition editor
9. Double click on "Click here to pick the part..." and select `[partial-id="/CompositionProvider_TestHelper/views/MainPage.html"]`
10. You are now editing the composition. Add something like `abc` to the composition
11. Press "Save"

You can repeat steps 6-11 to reproduce it again and again.

## Expected

It should work and the editor should contain:

```html
<style>:host{display:block}</style>
        <slot name="CompositionProvider_TestHelper/message"></slot>abc
    <slot></slot>
```

## Actual

It works but the editor contains redundant HTML:

```html
<style>:host{display:block}</style>
        <slot name="CompositionProvider_TestHelper/message"></slot>abc
    <slot></slot>
        <slot name="CompositionProvider_TestHelper/message"></slot>
``

**Comment by alshakero** [comment]
Okay so I have a theory of what's going on; most probably the issue comes from this loop:
https://github.com/Starcounter/starcounter-include/blob/a2663d12d43e8dc5d335e7642d9f25b6b78a5366/starcounter-include.html#L269-L280

More exactly is more from this line https://github.com/Starcounter/starcounter-include/blob/a2663d12d43e8dc5d335e7642d9f25b6b78a5366/starcounter-include.html#L275.

My theory is that we're creating `merged` outside the loop **once**. And then looping **several times** over every key. And this creates duplicates. 

Weirdly, even though this code looks equivalent to the original.
Replacing:
```js
this.template.scopedNodes.forEach((nodes) => {
    if(nodes.scope === key) {
        nodes.forEach((node) => appendComposition(merged, node));
    }
});
```
With
```js
this.template.scopedNodes.filter(n => n.scope === key).forEach(node =>
    appendComposition(merged, node));
```
... fixes the issue.

I'll look at again in the morning for a better understanding, I might be just tired to see the difference between those snippets.

**Comment by ubbeK** [comment]
> TeamCity uses MSBuild ToolsVersion 14.0 while should uses 15.0. @ubbeK could you please fix the CompositionProvider build settings?

Sorry for missing this. 

I changed the following

- MSBuild version: `Microsoft Build Tools 2015` => `Microsoft Build Tools 2017`
- MSBuild ToolsVersion: `14.0` => `15.0`

If it still does not work, please see picture below in order to change it, @un-tone now as admin rights for all TeamCity Apps builds.

compositionprovider

**Comment by un-tone** [comment]
Thank you @ubbeK. It works well.

**Comment by warpech** [comment]
Current status: @alshakero investigating https://github.com/Starcounter/CompositionProvider/pull/15#issuecomment-342509123.

**Comment by alshakero** [comment]
@warpech I think I know what's going on. The array [`ViewUris`](https://github.com/Starcounter/starcounter-include/blob/subset-composition/starcounter-include.html#L267) is what's is being used for deduping templates, right? We're searching for each partial URL, making sure it *isn't* there, and then appending it to the composition.

But what if `ViewUris` is always empty? Which is the case. You'll just continue appending and appending. And that's what's happening. A PoC would be to change [this line](https://github.com/Starcounter/starcounter-include/blob/subset-composition/starcounter-include.html#L267):

From `if (compositionProvider.ViewUris) {`

To `if (compositionProvider.ViewUris.length) {`

This fixes the bug without breaking `compositionProvider` tests.

===

But I assume my fix isn't an actual fix. I think there is a bug in `compositionProvider` (not sure); it's not updating `ViewUris` array.


**Potentially useful information:** when you run the tests `ViewUris` array does get populated.

**Comment by warpech** [comment]
@un-tone I need your help with this:

1, The code at https://github.com/Starcounter/CompositionProvider/blob/e6333ce9d4dbe73cb5a251a98af6ae42ddadc839/src/CompositionProvider/Helpers/HtmlViewCompositionHandler.cs#L110-L115 should be called also after a new instance of `HtmlViewComposition` is created by the `Update` function.
2. The commit https://github.com/Starcounter/CompositionProvider/pull/15/commits/f067c82c3070f71ac2d29e12316445b9c79b3d81 introduces old API compatibility but in a different fashion than I imagined. The old API will be removed in Starcounter 2.4, so I think it would be better if CompositionProvider upgrades the data from the old to the new API on startup. Then, all middleware and views CompositionProvider should only work on the new API.

**Comment by warpech** [comment]
Thanks @un-tone, for me this looks ready for final testing now!

@un-tone, @PatrykSzwer, @alshakero could you please help me to find final flaws. Preparation steps:

1. Use Starcounter 2.3.1.7764
2. Clone starcounter-include `subset-composition` branch and use `starcounter-include.html` from that branch. I am using Fiddler for that (see above for config), but you can also just copy `starcounter-include.html` to the Starcounter installation directory.
3. Clone, build and run CompositionEditor `develop` branch using `run.bat` (this will start CompositionEditor)
4. Clone, build and run CompositionProvider `layout-subset` branch using `test.bat` (this will start CompositionProvider, CompositionProvider_TestApp1, CompositionProvider_TestApp2, CompositionProvider_TestHelper)

Now that your set up is prepared, run **Test 1**, **Test 2** and **Test 3** which are explained in comments in this PR.

**Comment by un-tone** [comment]
Works perfect for me.

**Comment by PatrykSzwer** [comment]
Test 1: ✔️,
Test 2: ✔️,
Test 3: ✔️

File: README.md [hunk 1]
```diff
@@ -16,37 +18,43 @@ This file uses the following definitions:
...
 
 1. Check out [Starcounter/CompositionProvider](https://github.com/Starcounter/CompositionProvider) repo from GitHub
-2. Open `CompositionProvider.proj` in Visual Studio
-3. Build the solution
-4. Run it by launching `run.bat` in Windows, or executing `cmd //c "run.bat"` in Git Bash, or using Debug in Visual Studio
-5. Go to [http://localhost:8080/](http://localhost:8080/)
+1. Open `CompositionProvider.sln` in Visual Studio
+1. Build the solution
+1. Run it by launching `run.bat` in Windows, or executing `cmd //c "run.bat"` in Git Bash, or using Debug in Visual Studio
+1. Go to <http://localhost:8080/>
```
Comment (warpech) [inline README.md]: I didn't know this syntax, cool!

**Comment by warpech** [comment]
It looks ok to me too. @un-tone could you please merge it to develop? 

It still requires https://github.com/Starcounter/starcounter-include/pull/67

I suppose @Joltrix can test it after release.

**Comment by alshakero** [comment]
Test 1, 2 and 3 worked for me 🎊. Maybe @Joltrix or anyone will need clearer steps for Test 1 you can look at this GIF by @un-tone 
test1

**Pull request merged**